vip-api/src/main/java/fr/insalyon/creatis/vip/api/controller/admin/AppVersionController.java

```diff
@@ -0,0 +1,176 @@
+import fr.insalyon.creatis.vip.application.client.bean.Application;
+import fr.insalyon.creatis.vip.application.server.business.AppVersionBusiness;
+import fr.insalyon.creatis.vip.application.server.business.ApplicationBusiness;
+import fr.insalyon.creatis.vip.core.server.business.BusinessException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestBody;
+import jakarta.validation.Valid;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
+@RequestMapping("admin/appVersions")
+public class AppVersionController extends ApiController {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final ApplicationBusiness applicationBusiness;
+    private final AppVersionBusiness appVersionBusiness;
+
+    @Autowired
+    protected AppVersionController(ApplicationBusiness applicationBusiness, AppVersionBusiness appVersionBusiness) {
+        this.applicationBusiness = applicationBusiness;
+        this.appVersionBusiness = appVersionBusiness;
+    }
+
+    @RequestMapping(method = RequestMethod.GET)
+    public List<AppVersion> listAppVersions() throws ApiException {
+        logMethodInvocation(logger, "listAppVersions");
+        try {
+            List<Application> apps = applicationBusiness.getApplications();
+            List<AppVersion> appVersions = new ArrayList<>();
+            for (Application app : apps) {
+                appVersions.addAll(appVersionBusiness.getVersions(app.getName()));
+            }
+            return appVersions;
+        } catch (BusinessException e) {
+            throw new ApiException(e);
+        }
+    }
+
+    private class AppVersionStrings {
+        public String appName;
+        public String version;
+        AppVersionStrings(String appName, String version) {
+            this.appName = appName;
+            this.version = version;
+        }
+    }
+
+    private AppVersionStrings parseAppVersionId(String appVersionId) throws ApiException {
+        try {
+            int delimiterPos = appVersionId.lastIndexOf("/");
+            if (delimiterPos >= 0) {
+                String appName = appVersionId.substring(0, delimiterPos);
+                String version = appVersionId.substring(delimiterPos + 1);
+                return new AppVersionStrings(appName, version);
+            } else {
+                throw new ApiException("Invalid appVersionId");
+            }
+        } catch (ApiException e) {
+            logger.error("Error decoding appVersionId {}", appVersionId, e);
+            throw new ApiException("cannot decode appVersionId : " + appVersionId);
```

Comment:
is the exception raised in line 100 and catched just bellow or am i missing something ?

Reply:
You can reuse existing error codes : `throw new ApiException(ApiError.INVALID_PIPELINE_IDENTIFIER, identifier);`